ouster-ros/src/thread_safe_ring_buffer.h

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright (c) 2018-2023, Ouster, Inc.
+ * All rights reserved.
+ *
+ * @file thread_safe_ring_buffer.h
+ * @brief File contains thread safe implementation of a ring buffer
+ */
+
+#pragma once
+
+#include <iostream>
+#include <vector>
+#include <mutex>
+#include <condition_variable>
+
+/**
+ * @class ThreadSafeRingBuffer thread safe ring buffer.
+ */
+class ThreadSafeRingBuffer {
+public:
+    ThreadSafeRingBuffer(size_t item_size_, size_t items_count_) :
+        buffer(item_size_ * items_count_), item_size(item_size_),
+        max_items_count(items_count_), active_items_count(0),
+        write_idx(0), read_idx(0) {
+    }
+
+    /**
+     * Gets the maximum number of items that this ring buffer can hold.
+     */
+    size_t capacity() const { return max_items_count; }
+
+    /**
+     * Gets the number of item that currently occupy the ring buffer.
+     */
+    size_t size() const {
+        std::lock_guard<std::mutex> lock(mutex);
+        return active_items_count;
+    }
+
+    /**
+     * Checks if the ring buffer is empty.
+     */
+    bool empty() const {
+        std::lock_guard<std::mutex> lock(mutex);
+        return active_items_count == 0;
+    }
+
+    /**
+     * Checks if the ring buffer is full.
+     */
+    bool full() const {
+        std::lock_guard<std::mutex> lock(mutex);
+        return active_items_count == max_items_count;
+    }
+
+    /**
+     * Writes to the buffer safely, the method will keep blocking until the there
+     * is a space available within the buffer.
+     */
+    template <class BufferWriteT>
+    void write(BufferWriteT buffer_write) {
+        std::unique_lock<std::mutex> lock(mutex);
+        fullCondition.wait(lock, [this] { return active_items_count < capacity(); });
+        buffer_write(&buffer[write_idx * item_size]);
+        write_idx = (write_idx + 1) % capacity();
+        ++active_items_count;
+        emptyCondition.notify_one();
+    }
+
+
+    /**
+     * Writes to the buffer safely, if there is not space left then this method
+     * will overite the last item.
+     */
+    template <class BufferWriteT>
+    void write_overwrite(BufferWriteT buffer_write) {
+        std::unique_lock<std::mutex> lock(mutex);
+        buffer_write(&buffer[write_idx * item_size]);
+        write_idx = (write_idx + 1) % capacity();
+        if (active_items_count < capacity()) {
+            ++active_items_count;
+        } else {
+            read_idx = (read_idx + 1) % capacity();
+        }
+        emptyCondition.notify_one();
+    }
+
+    /**
+     * Gives access to read the buffer through a callback, the method will block
+     * until there is something to read is available.
+     */
+    template <typename BufferReadT>
+    void read(BufferReadT&& buffer_read) {
```

Comment:
BufferReadT is taken as universal reference vs getting copied as BufferWriteT above.
In some edge scenarios these may well be different, it's a good idea to keep these consistent unless we have a good reason to.

Reply:
On that note, do we have a convention for templated functors, like `BufferReadFn` instead of `BufferReadT`?

Reply:
Thanks for noting that, my intention was to accept as a universal reference to avoid copying the lambda functions when possible but I don't think I am currently making use of it. I will change update all the definitions to use &&.

I don't believe we have a convention for naming template typenames but using `BufferReadFn` seems appropriate.